### ISSUE TITLE
"Duplicate call to publish" error since 1.11.0-RC1

### ISSUE THREAD

**Issue by pirgeo** [micrometer-metrics/micrometer#3832]
**Describe the bug**
As part of some internal testing with our Dynatrace meter registry, I recently saw the following error message for the first time: 
```
Publishing is already in progress. Skipping duplicate call to publish().
```
This log message was introduced in https://github.com/micrometer-metrics/micrometer/pull/3714 and first released with Micrometer 1.11.0-RC1. 

I was trying to debug this a little bit, but it's really hard to reproduce (which is a good thing, I guess 😄). One thing that I did find out was that some change in 1.11.0-RC1 seems to lead to multiple publishing threads being created. E.g., when running the registry with micrometer core version 1.9.10 (the default for Spring Boot 2.7.11), I get 1 thread named `dynatrace-metrics-publisher`, which is responsible for pushing metrics every X seconds. The thread is created using the [default thread factory](https://github.com/micrometer-metrics/micrometer/blob/main/implementations/micrometer-registry-dynatrace/src/main/java/io/micrometer/dynatrace/DynatraceMeterRegistry.java#LL59C75-L59C75), which is an instance of [NamedThreadFactory](https://github.com/micrometer-metrics/micrometer/blob/main/micrometer-core/src/main/java/io/micrometer/core/instrument/util/NamedThreadFactory.java). Everything is fine and seems to work in this version.

However, when I switch over to Micrometer 1.11.0-RC1, I suddenly have two publishing threads ( `dynatrace-metrics-publisher` and  `dynatrace-metrics-publisher-2`). It, therefore, seems as if the NewThread method on the ThreadFactory is called twice. I am not quite sure how this happens, since I assume the threads are created [here](https://github.com/micrometer-metrics/micrometer/blob/main/micrometer-core/src/main/java/io/micrometer/core/instrument/step/StepMeterRegistry.java#L118) and the SingleThreadScheduledExecutor should not create two threads (maybe?). 

I have also tried another registry (I randomly picked the elastic registry), and it seems to also be creating two export threads instead of one. Hence, I believe this might have something to do with the base class (maybe StepMeterRegistry?). Have you seen this before? Any ideas where it could come from? As I said, I am having a hard time getting a reproducer ready (getting two threads is easy, but they seem to not interfere most of the time). I assume the error comes from a race condition when both threads try to export at the same time. This would also explain why it only shows up once in a blue moon. 

**Environment**

 - Micrometer version: 1.11.0-RC1
 - Micrometer registry: I assume StepMeterRegistry, confirmed in DynatraceMeterRegistry and ElasticMeterRegistry, possibly others.
 - OS: WSL2 on Windows 11
 - Java version: `openjdk version "17.0.4" 2022-07-19`

**To Reproduce**
How to reproduce the bug: - 

**Expected behavior**
I expected to not see that error and to only have one live publishing thread.

**Comment by lenin-jaganathan** [comment]
Duplicate threads might be due to https://github.com/micrometer-metrics/micrometer/pull/3750. And this is an expected behaviour for StepRegistries to have a poll thread running. And only one will be responsible for publishing while other thread only takes care of rotating the StepMeters on Step crossover.

One possible cause for the message you are seeing is, if the publishing takes more than one step.

**Comment by pirgeo** [comment]
Could you expand on what you mean by "publishing takes more than one step"? Do you mean if it takes longer than the export interval to prepare the metrics? 

Also, I missed #3750 when it was created. I looked at it briefly - Do I get it right that there will be a separate thread calling `count` on most Meter types periodically in order to roll over finished time slots before `publish` is called? I think this should not interfere with our export (we have get-and-reset semantics in most of our Meters in the Dynatrace Registry, and calling count won't do anything, except for Counters, where we use the standard StepCounter). From what I understand right now, the change in #3750 should not be responsible for the error, thanks for pointing that out though!

**Comment by jonatan-ivanov** [comment]
Quick question because I would dive deeper to publishing: are you getting this log message when you shut down your app or it happens randomly way before you would stop your app?
At shutdown, the registry should be closed, that will trigger a final publish. If there is already a publish in-progress, you might get the log message above.

**Comment by lenin-jaganathan** [comment]
> Could you expand on what you mean by "publishing takes more than one step"? Do you mean if it takes longer than the export interval to prepare the metrics?

Yes, you are right. If the step is small let's say 10 seconds and the publishing takes more than that, then it is possible to get this message logged.

> From what I understand right now, the change in https://github.com/micrometer-metrics/micrometer/pull/3750 should not be responsible for the error, thanks for pointing that out though!
Yes. This will not cause any issues, but this will answer your question on two publishing threads ( dynatrace-metrics-publisher and dynatrace-metrics-publisher-2). (This is well described in https://github.com/micrometer-metrics/micrometer/issues/3837)

**Comment by lenin-jaganathan** [comment]
> I think this should not interfere with our export (we have get-and-reset semantics in most of our Meters in the Dynatrace Registry, and calling count won't do anything, except for Counters, where we use the standard StepCounter)

Just looked at the Dynatrace implementation. And it seems, it is a bit tricky with regards to this poll service and how it is implemented in Dynatrace. There are some key takeaways,
* DynaTrace is a step registry but uses AbstractTimers and AbstractDistributionSummary. This means the poll service will roll the counters and not the Timers/Ds. So, you might see a behavior where the counters report values for absolute steps 00:00:00-00:00:01 but Timer/Ds reporting data whenever they are queried (which might be anywhere from 0-80% of step duration. If the step is 60seconds, then exporting can **start** anytime between 0-48 seconds). So, this should be revisited if counters and Timers/DS reporting for different durations is the right approach.

Didn't want to diverge from the issue, but wanted to bring this to your attention.

**Comment by pirgeo** [comment]
> Quick question before I would dive deeper to publishing: are you getting this log message when you shut down your app or it happens randomly way before you would stop your app?

@jonatan-ivanov I checked, and it was actually the case that the app was shutting down. I had called the actuator shutdown endpoint, and the logs appeared after the shutdown call, so I think your explanation makes sense here. See here for full logs with annotations:

<details>

```
[2023-05-09T15:32:58.211Z] 
[2023-05-09T15:32:58.211Z]   .   ____          _            __ _ _
[2023-05-09T15:32:58.211Z]  /\\ / ___'_ __ _ _(_)_ __  __ _ \ \ \ \
[2023-05-09T15:32:58.211Z] ( ( )\___ | '_ | '_| | '_ \/ _` | \ \ \ \
[2023-05-09T15:32:58.211Z]  \\/  ___)| |_)| | | | | || (_| |  ) ) ) )
[2023-05-09T15:32:58.211Z]   '  |____| .__|_| |_|_| |_\__, | / / / /
[2023-05-09T15:32:58.211Z]  =========|_|==============|___/=/_/_/_/
[2023-05-09T15:32:58.211Z]  :: Spring Boot ::            (v3.1.0-RC1)
[2023-05-09T15:32:58.211Z] 
[2023-05-09T15:32:58.211Z] 2023-05-09T15:32:57.941Z  INFO 2026 --- [           main] c.d.m.t.d.MyApp             : Starting MyApp using Java 17.0.3 with PID 2026 (***)
[2023-05-09T15:32:58.211Z] 2023-05-09T15:32:57.945Z  INFO 2026 --- [           main] c.d.m.t.d.MyApp             : The following 1 profile is active: "dynatrace"
[2023-05-09T15:33:01.310Z] 2023-05-09T15:33:01.008Z  INFO 2026 --- [           main] o.s.b.w.embedded.tomcat.TomcatWebServer  : Tomcat initialized with port(s): 8080 (http)
[2023-05-09T15:33:01.310Z] 2023-05-09T15:33:01.030Z  INFO 2026 --- [           main] o.apache.catalina.core.StandardService   : Starting service [Tomcat]
[2023-05-09T15:33:01.310Z] 2023-05-09T15:33:01.031Z  INFO 2026 --- [           main] o.apache.catalina.core.StandardEngine    : Starting Servlet engine: [Apache Tomcat/10.1.8]
[2023-05-09T15:33:01.711Z] 2023-05-09T15:33:01.430Z  INFO 2026 --- [           main] o.a.c.c.C.[Tomcat].[localhost].[/]       : Initializing Spring embedded WebApplicationContext
[2023-05-09T15:33:01.712Z] 2023-05-09T15:33:01.432Z  INFO 2026 --- [           main] w.s.c.ServletWebServerApplicationContext : Root WebApplicationContext: initialization completed in 3301 ms
[2023-05-09T15:33:01.712Z] 2023-05-09T15:33:01.611Z  INFO 2026 --- [           main] i.m.dynatrace.DynatraceMeterRegistry     : Exporting to Dynatrace metrics API v2
[2023-05-09T15:33:01.712Z] 2023-05-09T15:33:01.614Z  INFO 2026 --- [           main] i.m.dynatrace.v2.DynatraceExporterV2     : Exporting to endpoint ****
[2023-05-09T15:33:01.712Z] 2023-05-09T15:33:01.623Z  INFO 2026 --- [           main] i.m.c.instrument.push.PushMeterRegistry  : publishing metrics for DynatraceMeterRegistry every 1m
[2023-05-09T15:33:03.417Z] 2023-05-09T15:33:03.316Z  INFO 2026 --- [           main] o.s.b.a.e.web.EndpointLinksResolver      : Exposing 14 endpoint(s) beneath base path '/actuator'
[2023-05-09T15:33:03.808Z] 2023-05-09T15:33:03.450Z  INFO 2026 --- [           main] o.s.b.w.embedded.tomcat.TomcatWebServer  : Tomcat started on port(s): 8080 (http) with context path ''
[2023-05-09T15:33:03.808Z] 2023-05-09T15:33:03.527Z  INFO 2026 --- [           main] c.d.m.t.d.MyApp             : Started MyApp in 6.71 seconds (process running for 7.802)
[2023-05-09T15:33:04.199Z] 2023-05-09T15:33:03.917Z  INFO 2026 --- [nio-8080-exec-1] o.a.c.c.C.[Tomcat].[localhost].[/]       : Initializing Spring DispatcherServlet 'dispatcherServlet'
[2023-05-09T15:33:04.199Z] 2023-05-09T15:33:03.918Z  INFO 2026 --- [nio-8080-exec-1] o.s.web.servlet.DispatcherServlet        : Initializing Servlet 'dispatcherServlet'
[2023-05-09T15:33:04.199Z] 2023-05-09T15:33:03.920Z  INFO 2026 --- [nio-8080-exec-1] o.s.web.servlet.DispatcherServlet        : Completed initialization in 1 ms

// At this point here I ping the health check endpoint and make sure the Spring Boot app is up and running, then I call some of the APIs exposed by the web app. 
// Calling the endpoints will create metrics which will be forwarded to Dynatrace. Metrics are exported:

[2023-05-09T15:34:44.274Z] 2023-05-09T15:34:34.101Z DEBUG 2026 --- [trics-publisher] i.m.dynatrace.v2.DynatraceExporterV2     : Sending 2 lines to ****
[2023-05-09T15:34:44.275Z] 2023-05-09T15:34:34.101Z DEBUG 2026 --- [trics-publisher] i.m.dynatrace.v2.DynatraceExporterV2     : Sending lines:
[2023-05-09T15:34:44.275Z] micrometer.custom.gauge,dt.metrics.source=micrometer,example_key=example_value gauge,5.234 1683646474093
[2023-05-09T15:34:44.275Z] micrometer.custom.timer,dt.metrics.source=micrometer,example_key=example_value gauge,min=23.0,max=23.0,sum=23.0,count=1 1683646474096

// Then I ping the actuator shutdown endpoint to shut down the Spring Boot app.

[2023-05-09T15:34:44.275Z] 2023-05-09T15:34:34.734Z  INFO 2026 --- [       Thread-2] o.apache.catalina.core.StandardService   : Stopping service [Tomcat]
[2023-05-09T15:34:44.275Z] 2023-05-09T15:34:34.737Z  INFO 2026 --- [       Thread-2] o.a.c.c.C.[Tomcat].[localhost].[/]       : Destroying Spring FrameworkServlet 'dispatcherServlet'
[2023-05-09T15:34:44.275Z] 2023-05-09T15:34:34.741Z  WARN 2026 --- [       Thread-2] o.a.c.loader.WebappClassLoaderBase       : The web application [ROOT] appears to have started a thread named [dynatrace-metrics-publisher] but has failed to stop it. This is very likely to create a memory leak. Stack trace of thread:
[2023-05-09T15:34:44.275Z]  java.base@17.0.3/sun.security.util.math.intpoly.IntegerPolynomial.conditionalSwap(IntegerPolynomial.java:463)
[2023-05-09T15:34:44.275Z]  java.base@17.0.3/sun.security.util.math.intpoly.IntegerPolynomial$MutableElement.conditionalSwapWith(IntegerPolynomial.java:650)
[2023-05-09T15:34:44.275Z]  platform/jdk.crypto.ec@17.0.3/sun.security.ec.XECOperations.cswap(XECOperations.java:183)
[2023-05-09T15:34:44.275Z]  platform/jdk.crypto.ec@17.0.3/sun.security.ec.XECOperations.pointMultiply(XECOperations.java:230)
[2023-05-09T15:34:44.275Z]  platform/jdk.crypto.ec@17.0.3/sun.security.ec.XECOperations.computePublic(XECOperations.java:77)
[2023-05-09T15:34:44.275Z]  platform/jdk.crypto.ec@17.0.3/sun.security.ec.XDHKeyPairGenerator.generateKeyPair(XDHKeyPairGenerator.java:109)
[2023-05-09T15:34:44.275Z]  java.base@17.0.3/java.security.KeyPairGenerator$Delegate.generateKeyPair(KeyPairGenerator.java:727)
[2023-05-09T15:34:44.275Z]  java.base@17.0.3/sun.security.ssl.XDHKeyExchange$XDHEPossession.<init>(XDHKeyExchange.java:105)
[2023-05-09T15:34:44.275Z]  java.base@17.0.3/sun.security.ssl.NamedGroup$XDHScheme.createPossession(NamedGroup.java:611)
[2023-05-09T15:34:44.275Z]  java.base@17.0.3/sun.security.ssl.NamedGroup$NamedGroupSpec.createPossession(NamedGroup.java:520)
[2023-05-09T15:34:44.275Z]  java.base@17.0.3/sun.security.ssl.NamedGroup.createPossession(NamedGroup.java:428)
[2023-05-09T15:34:44.275Z]  java.base@17.0.3/sun.security.ssl.SSLKeyExchange$T13KeyAgreement.createPossession(SSLKeyExchange.java:588)
[2023-05-09T15:34:44.275Z]  java.base@17.0.3/sun.security.ssl.SSLKeyExchange.createPossessions(SSLKeyExchange.java:97)
[2023-05-09T15:34:44.275Z]  java.base@17.0.3/sun.security.ssl.KeyShareExtension$CHKeyShareProducer.getShare(KeyShareExtension.java:299)
[2023-05-09T15:34:44.275Z]  java.base@17.0.3/sun.security.ssl.KeyShareExtension$CHKeyShareProducer.produce(KeyShareExtension.java:261)
[2023-05-09T15:34:44.275Z]  java.base@17.0.3/sun.security.ssl.SSLExtension.produce(SSLExtension.java:599)
[2023-05-09T15:34:44.275Z]  java.base@17.0.3/sun.security.ssl.SSLExtensions.produce(SSLExtensions.java:265)
[2023-05-09T15:34:44.275Z]  java.base@17.0.3/sun.security.ssl.ClientHello$ClientHelloKickstartProducer.produce(ClientHello.java:638)
[2023-05-09T15:34:44.275Z]  java.base@17.0.3/sun.security.ssl.SSLHandshake.kickstart(SSLHandshake.java:529)
[2023-05-09T15:34:44.275Z]  java.base@17.0.3/sun.security.ssl.ClientHandshakeContext.kickstart(ClientHandshakeContext.java:112)
[2023-05-09T15:34:44.275Z]  java.base@17.0.3/sun.security.ssl.TransportContext.kickstart(TransportContext.java:251)
[2023-05-09T15:34:44.275Z]  java.base@17.0.3/sun.security.ssl.SSLSocketImpl.startHandshake(SSLSocketImpl.java:448)
[2023-05-09T15:34:44.275Z]  java.base@17.0.3/sun.security.ssl.SSLSocketImpl.startHandshake(SSLSocketImpl.java:426)
[2023-05-09T15:34:44.275Z]  java.base@17.0.3/sun.net.www.protocol.https.HttpsClient.afterConnect(HttpsClient.java:580)
[2023-05-09T15:34:44.275Z]  java.base@17.0.3/sun.net.www.protocol.https.AbstractDelegateHttpsURLConnection.connect(AbstractDelegateHttpsURLConnection.java:183)
[2023-05-09T15:34:44.275Z]  java.base@17.0.3/sun.net.www.protocol.http.HttpURLConnection.getOutputStream0(HttpURLConnection.java:1430)
[2023-05-09T15:34:44.275Z]  java.base@17.0.3/sun.net.www.protocol.http.HttpURLConnection.getOutputStream(HttpURLConnection.java:1401)
[2023-05-09T15:34:44.275Z]  java.base@17.0.3/sun.net.www.protocol.https.HttpsURLConnectionImpl.getOutputStream(HttpsURLConnectionImpl.java:220)
[2023-05-09T15:34:44.275Z]  app//io.micrometer.core.ipc.http.HttpUrlConnectionSender.send(HttpUrlConnectionSender.java:98)
[2023-05-09T15:34:44.275Z]  app//io.micrometer.core.ipc.http.HttpSender$Request$Builder.send(HttpSender.java:306)
[2023-05-09T15:34:44.275Z]  app//io.micrometer.dynatrace.v2.DynatraceExporterV2.send(DynatraceExporterV2.java:331)
[2023-05-09T15:34:44.275Z]  app//io.micrometer.dynatrace.v2.DynatraceExporterV2$$Lambda$1014/0x00000008010aa730.accept(Unknown Source)
[2023-05-09T15:34:44.275Z]  java.base@17.0.3/java.lang.Iterable.forEach(Iterable.java:75)
[2023-05-09T15:34:44.275Z]  app//io.micrometer.dynatrace.v2.DynatraceExporterV2.sendInBatches(DynatraceExporterV2.java:374)
[2023-05-09T15:34:44.275Z]  app//io.micrometer.dynatrace.v2.DynatraceExporterV2.export(DynatraceExporterV2.java:144)
[2023-05-09T15:34:44.275Z]  app//io.micrometer.dynatrace.DynatraceMeterRegistry.publish(DynatraceMeterRegistry.java:100)
[2023-05-09T15:34:44.275Z]  app//io.micrometer.core.instrument.push.PushMeterRegistry.publishSafely(PushMeterRegistry.java:64)
[2023-05-09T15:34:44.276Z]  app//io.micrometer.core.instrument.push.PushMeterRegistry$$Lambda$618/0x0000000800f04750.run(Unknown Source)
[2023-05-09T15:34:44.276Z]  java.base@17.0.3/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
[2023-05-09T15:34:44.276Z]  java.base@17.0.3/java.util.concurrent.FutureTask.runAndReset(FutureTask.java:305)
[2023-05-09T15:34:44.276Z]  java.base@17.0.3/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:305)
[2023-05-09T15:34:44.276Z]  java.base@17.0.3/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
[2023-05-09T15:34:44.276Z]  java.base@17.0.3/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
[2023-05-09T15:34:44.276Z]  java.base@17.0.3/java.lang.Thread.run(Thread.java:833)
[2023-05-09T15:34:44.276Z] 2023-05-09T15:34:34.742Z  WARN 2026 --- [       Thread-2] o.a.c.loader.WebappClassLoaderBase       : The web application [ROOT] appears to have started a thread named [dynatrace-metrics-publisher-2] but has failed to stop it. This is very likely to create a memory leak. Stack trace of thread:
[2023-05-09T15:34:44.276Z]  java.base@17.0.3/jdk.internal.misc.Unsafe.park(Native Method)
[2023-05-09T15:34:44.276Z]  java.base@17.0.3/java.util.concurrent.locks.LockSupport.parkNanos(LockSupport.java:252)
[2023-05-09T15:34:44.276Z]  java.base@17.0.3/java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.awaitNanos(AbstractQueuedSynchronizer.java:1672)
[2023-05-09T15:34:44.276Z]  java.base@17.0.3/java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue.take(ScheduledThreadPoolExecutor.java:1182)
[2023-05-09T15:34:44.276Z]  java.base@17.0.3/java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue.take(ScheduledThreadPoolExecutor.java:899)
[2023-05-09T15:34:44.276Z]  java.base@17.0.3/java.util.concurrent.ThreadPoolExecutor.getTask(ThreadPoolExecutor.java:1062)
[2023-05-09T15:34:44.276Z]  java.base@17.0.3/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1122)
[2023-05-09T15:34:44.276Z]  java.base@17.0.3/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
[2023-05-09T15:34:44.276Z]  java.base@17.0.3/java.lang.Thread.run(Thread.java:833)
[2023-05-09T15:34:44.276Z] 2023-05-09T15:34:34.905Z  WARN 2026 --- [       Thread-2] i.m.c.instrument.push.PushMeterRegistry  : Publishing is already in progress. Skipping duplicate call to publish().
[2023-05-09T15:34:44.276Z] 

// Note that no further metrics are exported anymore. This would be Debug logged, so it seems the two threads are blocking each other from exporting or so.
// The final export therefore never happens, and the data does not show up in Dynatrace.
```

</details>

> Just looked at the Dynatrace implementation. And it seems, it is a bit tricky with regards to this poll service and how it is implemented in Dynatrace.

@lenin-jaganathan, yes, thank you! I realized too late that there was such a change and I will have to think about how to address this. I guess for 1.11.0 we will have the mixed version. I am looking into having a uniform solution again.

**Comment by lenin-jaganathan** [comment]
> Note that no further metrics are exported anymore. This would be Debug logged, so it seems the two threads are blocking each other from exporting or so.

Can you elaborate on this? I don't expect threads to be blocked. Not really able to get the point🙂.

> The final export therefore never happens, and the data does not show up in Dynatrace.

I guess if the publishing is already in progress when the app closes, then it is the final publishing right? Am I missing something?

**Comment by lenin-jaganathan** [comment]
> but does not export the last value.

Still confusing here. The publishing will log the message only if there is another publishing is going on. And by that, you should get the accumulated values. Am I missing something??

**Comment by jonatan-ivanov** [comment]
After reading through the comments:

Getting this message is completely normal:
```
Publishing is already in progress. Skipping duplicate call to publish().
```
It just means that you shut down your app in the middle of a publish so Micrometer did not do an extra final publish triggered by closing the registry because it was already doing the final publish.

Or at least that's how it suppose to work, but this might be concerning (we might have a bug):
>The final export therefore never happens, and the data does not show up in Dynatrace.

@pirgeo Does this mean that no publish happen around application shut down or only one happens (and you would expect two)?

For the record, these are valid scenarios:
1. "Normal" publish + "final" publish (2): publishing happens triggered by the scheduler, then the shutdown process starts an extra (off-cycle) publish triggered by calling `close` (~you see two publish in one step)
2. "Final" publish is the "normal" publish (1): the shutdown process wants to start an extra (off-cycle) publish triggered by calling `close` but there is already a publish in-progress triggered by the scheduler so this in-progress publish is your final one, Micrometer does not do another one

Based on your logs it seems the second scenario is happening. Are you losing data because traffic was still flowing in between the start of the scheduled publishing and Tomcat stopped receiving new requests? Ideally, traffic should have been already stopped by the time the app gets the shutdown signal.

**Comment by pirgeo** [comment]
Hi, I looked into this a little more. I think it's closer to the second one, but there is something else going on: 
```
// PushMeterRegistry is started at 15:33:01
[2023-05-09T15:33:01.712Z] 2023-05-09T15:33:01.623Z  INFO 2026 --- [           main] i.m.c.instrument.push.PushMeterRegistry  : publishing metrics for DynatraceMeterRegistry every 1m
// Pinging the endpoint happened at 15:33:04 (I had excluded that log line initially)
[2023-05-09T15:33:04.589Z] got response for http://localhost:8080/api/timer - {Instrument:timer Value:23 StatusCode:200}
// First export happens at 15:34:44
[2023-05-09T15:34:44.274Z] 2023-05-09T15:34:34.101Z DEBUG 2026 --- [trics-publisher] i.m.dynatrace.v2.DynatraceExporterV2     : Sending 2 lines to <my endpoint>
// Hitting the actuator shutdown endpoint at 15:34:44
[2023-05-09T15:34:44.275Z] 2023/05/09 15:34:34 attempting to shut down Spring Boot app via shutdown endpoint
[2023-05-09T15:34:44.275Z] 2023-05-09T15:34:34.734Z  INFO 2026 --- [       Thread-2] o.apache.catalina.core.StandardService   : Stopping service [Tomcat]
// error message at 15:34:44
[2023-05-09T15:34:44.276Z] 2023-05-09T15:34:34.905Z  WARN 2026 --- [       Thread-2] i.m.c.instrument.push.PushMeterRegistry  : Publishing is already in progress. Skipping duplicate call to publish().
```
Interestingly, in the example above, I never see a log line like: 
```
...  i.m.dynatrace.v2.DynatraceExporterV2     : Sent 2 metric lines, linesOk: 2, linesInvalid: 0.
```
which would indicate that the request completed successfully. This would align with your assumption that Tomcat is already down when the response arrives (although I thought that Tomcat was responsible for handling the requests that go to the app, not the requests that Micrometer sends out). Either way, it seems the request does not complete, and the data does not show up. 

The app was running from 15:32:58 to 15:34:44, so roughly 1 minute and 46 seconds. At 15:33:01, the exporter thread is started. At 15:33:04, an instrument is created. At 15:34:44 (so 1:40 after the start of the exporter thread) the app is shut down. There was only one (attempted) export, which coincides exactly with the shutdown of the app. I would have expected two exports: one at around 15:34:01, about a minute after the exporter thread starts, and one on shutdown. It seems that both exports are happening at the same time (hence the duplicate publishing error). Even with the error, one of the threads should successfully export the data, yet the "Sent 2 metric lines, linesOk: 2, linesInvalid: 0." line is never printed. I tried reproducing it, and usually, when I hit the shutdown endpoint, it will tell me: 
```
[2023-05-23T14:51:21.884Z] 2023-05-23T14:51:21.431Z  INFO 2225 --- [       Thread-2] o.apache.catalina.core.StandardService   : Stopping service [Tomcat]
[2023-05-23T14:51:21.884Z] 2023-05-23T14:51:21.435Z  INFO 2225 --- [       Thread-2] o.a.c.c.C.[Tomcat].[localhost].[/]       : Destroying Spring FrameworkServlet 'dispatcherServlet'
[2023-05-23T14:51:21.884Z] 2023-05-23T14:51:21.441Z  WARN 2225 --- [       Thread-2] o.a.c.loader.WebappClassLoaderBase       : The web application [ROOT] appears to have started a thread named [dynatrace-metrics-publisher] but has failed to stop it. This is very likely to create a memory leak. Stack trace of thread:
...
[2023-05-23T14:51:21.884Z] 2023-05-23T14:51:21.442Z  WARN 2225 --- [       Thread-2] o.a.c.loader.WebappClassLoaderBase       : The web application [ROOT] appears to have started a thread named [dynatrace-metrics-publisher-2] but has failed to stop it. This is very likely to create a memory leak. Stack trace of thread:
...
[2023-05-23T14:51:21.884Z] 2023-05-23T14:51:21.455Z DEBUG 2225 --- [       Thread-2] i.m.dynatrace.v2.DynatraceExporterV2     : Sending 2 lines to <my endpoint>
...
[2023-05-23T14:51:22.334Z] 2023-05-23T14:51:21.786Z DEBUG 2225 --- [       Thread-2] i.m.dynatrace.v2.DynatraceExporterV2     : Sent 2 metric lines, linesOk: 2, linesInvalid: 0.
```

Note that in this case the "Sent 2 metric lines" message is there. It seems like there is something blocking inside that does not allow the request to finish in the case that two threads (the one triggered by shutdown and the one triggered by the clock) are exporting at the same time. 

**TL;DR**: It seems like when two threads are trying to export at the same time, the request does not complete.

This is a long shot, but is it possible that the app shuts down before the request completes because one of the threads reports "there is nothing to do, we can shut down" while the request is still active in the other thread?

**Comment by jonatan-ivanov** [comment]
>Interestingly, in the example above, I never see a log line like:
```
...  i.m.dynatrace.v2.DynatraceExporterV2     : Sent 2 metric lines, linesOk: 2, linesInvalid: 0.
```

You should see that line too except in one case I think: when we call `.shutdown()` on the `ScheduledExecutorService` in `PushMeterRegistry`, that method does not wait for previously submitted tasks to complete execution. This means that if you shut down in the middle of publishing the JVM will not wait for the publishing to happen, it can shutdown before the backend returns the response and you would see that log message.

>although I thought that Tomcat was responsible for handling the requests that go to the app, not the requests that Micrometer sends out

Sorry for the confusion, Tomcat has nothing to do with the communication between Micrometer and the backend. What I meant above is a different corner case but it seems it's not an issue for you. (Let me clarify anyways in case you are curious: let's say Micrometer is sending the data to the backend at this moment, and you still have traffic flowing in and you shut down the app. Those metrics that were recorded during publish or shortly after it will not be reported since we skip the second publish on close. Though now that Im thinking about this more, the second publish would have the same issue since traffic can still flow to your app during it, so I guess the only solution for this scenario is stopping traffic before shutdown.)

>Either way, it seems the request does not complete, and the data does not show up.

I guess this was because shutdown does not wait the publisher thread to finish, see above.

>I would have expected two exports: one at around 15:34:01, about a minute after the exporter thread starts, and one on shutdown.

This is a behavior we changed recently (see https://github.com/micrometer-metrics/micrometer/issues/2818). We used to publish at the same time after the step finished but if you have lots of apps and they all publish at the same time they can overwhelm the backend with bursting traffic. Because of this, publishing happens at a random time, if step is 60s, publishing can start anytime between 0-48s (0-80%) after the step has finished. I guess for you this was 44 seconds and it happened to be the same as shutdown and publish was slow so the JVM stopped before publish had a chance to finish.

>This is a long shot, but is it possible that the app shuts down before the request completes because one of the threads reports "there is nothing to do, we can shut down" while the request is still active in the other thread?

I think this is the case except you don't need two threads for this, the JVM simply does not wait for Micrometer to finish publishing. This is intentional (Micrometer should not disturb your app) but was already reported earlier, maybe we should discuss again if we should provide a config flag to wait for publish to finish (with a timeout).

**Comment by jonatan-ivanov** [comment]
>Let me see if I understand correctly...

I think you do. The global alignment is an issue if you have lots of instances and they publish at the exact same second.

>Would the JVM wait for the export triggered by pinging the Actuator shutdown endpoint? It does not wait for the export that is triggered by the timed thread, as you mentioned above.

This is intentional but I understand if this is concerning. Though this should be a different issue if it seems problematic.

>Finally, if my app ran longer than 2 minutes, I should not see that problem occur (since the first export cycle should definitely be done by 1:48 + the time it takes to export), is that right?

Yes. But if traffic keeps flowing and you keep sending requests after the first minute, the recorded data could have the same issue again if the JVM shuts down before the data is published.

**Comment by pirgeo** [comment]
> This is intentional but I understand if this is concerning. 

If it waits for the one triggered by a "manual" shutdown, that is okay. This will probably cover most cases (as long as people don't just kill their apps but actually shut them down properly). I agree, however, that it would be good to somehow be able to configure the shutdown to wait until the "timed" shutdown is also finished. Reliably exporting the last set of data points seems pretty important.

> But if traffic keeps flowing and you keep sending requests after the first minute, the recorded data could have the same issue again if the JVM shuts down before the data is published.

Yes, but only the data that comes in between the point in time where the export is started and the time that the export is finished, which should be less than a few seconds of data, if even that, correct?

**Comment by jonatan-ivanov** [comment]
>If it waits for the one triggered by a "manual" shutdown, that is okay.

Right now we don't wait, if there is a publish (either triggered by the scheduler or by close) and the JVM shuts down, we don't block the process. Let me discuss the option of adding an option for waiting with @shakuzen next week, I feel that this could be a valid argument now that we have publish on close but on the other hand Micrometer does everything it can to not disrupt the app.

>Yes, but only the data that comes in between the point in time where the export is started and the time that the export is finished, which should be less than a few seconds of data, if even that, correct?

Yes, if the publish can happen but if shutdown is fast enough and publish does not have time to do its job (like in your case), you will lose the data that was recorded since the start of the step.

**Comment by pirgeo** [comment]
> Let me discuss the option of adding an option for waiting... 

Great, thank you for taking this! I feel that in certain situations (where it is imperative that the last batch of data points arrives) users would want to ensure the delivery, even if it means waiting a few extra secods to shut down the app. It could be like a "ensure last metric batch delivery"-mode, that can be turned on if required. I would have to look closer, but this might then also make the Dynatrace get-and-reset semantics obsolete, if we can rely on getting the last batch out before shutting down.

> Yes, if the publish can happen but if shutdown is fast enough and publish does not have time to do its job (like in your case), you will lose the data that was recorded since the start of the step.

Got it 👍

**Comment by pirgeo** [comment]
So, since everything seems to be working as intended, I guess I can close this issue. My error was caused by the app not running long enough (e.g. the first export could be triggered at 1:48 min after the app starts, but my app only ran for 1:30 min). The duplicate call to publish was caused by the export being triggered by the timer at the same time as the shutdown was already in progress. Data was not exported since the JVM does not wait for the export process to finish. Should I open a separate issue about this? I would love to have a configurable timeout that the exporter is allowed to block the app from shutting down in order to finish the last export.

**Comment by shakuzen** [comment]
I suspect the block for publishing on shutdown behavior you want will likely be restored with the fix to #3846. It describes a different issue caused by the same underlying implementation, though. It's probably better to open a separate issue so we make sure we test properly for it and it doesn't slip through the cracks. Would you please open a new one and we can close this one?

**Comment by pirgeo** [comment]
@lenin-jaganathan @jonatan-ivanov @shakuzen Thank you very much for looking into this issue with me. I opened #3872 in order to track the potential addition of a timeout to wait for the final export to finish.
